sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala

```diff
@@ -0,0 +1,183 @@
+            errorClass = "INVALID_INDEX_COLUMN",
+            messageParameters = Map(
+              "reason" -> s"Index column must be an atomic attribute")
+          )
+      }
+
+      // Cast non-index columns to the least common type
+      val nonIndexColumnsAttr = child.output.filterNot(
+        _.exprId == indexColumn.asInstanceOf[Attribute].exprId)
+      val nonIndexTypes = nonIndexColumnsAttr.map(_.dataType)
+      val commonType = leastCommonType(nonIndexTypes)
+      val nonIndexColumnsAsLCT = nonIndexColumnsAttr.map { attr =>
+        Alias(Cast(attr, commonType), attr.name)()
+      }
+
+      // Exclude nulls and sort index column values, and collect the casted frame
+      val allCastCols = indexColumnAsString +: nonIndexColumnsAsLCT
+      val nonNullChild = Filter(IsNotNull(indexColumn), child)
+      val sortedChild = Sort(
+        Seq(SortOrder(indexColumn.asInstanceOf[Attribute], Ascending)),
+        global = true,
+        nonNullChild
+      )
+      val projectAllCastCols = Project(allCastCols, sortedChild)
+      val maxValues = sparkSession.sessionState.conf.dataFrameTransposeMaxValues
+      val limit = Literal(maxValues + 1)
+      val limitedProject = Limit(limit, projectAllCastCols)
+      val queryExecution = sparkSession.sessionState.executePlan(limitedProject)
+      val fullCollectedRows = queryExecution.executedPlan.executeCollect()
+
+      if (fullCollectedRows.isEmpty) {
+        // Return a DataFrame with a single column "key" containing non-index column names
+        val keyAttr = AttributeReference("key", StringType, nullable = false)()
+        val keyValues = nonIndexColumnsAttr.map(
+          _.name).map(name => UTF8String.fromString(name))
+        val keyRows = keyValues.map(value => InternalRow(value))
+
+        LocalRelation(Seq(keyAttr), keyRows)
+      } else {
+        val rowCount = fullCollectedRows.length
+        if (rowCount > maxValues) {
+          throw new AnalysisException(
+            errorClass = "EXCEED_ROW_LIMIT",
+            messageParameters = Map(
+              "maxValues" -> maxValues.toString,
+              "config" -> SQLConf.DATAFRAME_TRANSPOSE_MAX_VALUES.key))
+        }
+
+        // Transpose the matrix
+        val nonIndexColumnNames = nonIndexColumnsAttr.map(_.name)
+        val nonIndexColumnDataTypes = projectAllCastCols.output.tail.map(attr => attr.dataType)
+        val transposedMatrix = transposeMatrix(
+          fullCollectedRows, nonIndexColumnNames, nonIndexColumnDataTypes)
+        val transposedInternalRows = transposedMatrix.map { row =>
+          InternalRow.fromSeq(row.toIndexedSeq)
+        }
+
+        // Construct output attributes
+        val keyAttr = AttributeReference("key", StringType, nullable = false)()
+        val transposedColumnNames = fullCollectedRows.map { row => row.getString(0) }
+        val valueAttrs = transposedColumnNames.map { value =>
+          AttributeReference(
+            value,
+            commonType
```

Comment:
do we need to infer the nullability here?
e.g. if all non-index columns are nullable=false, then it seems the transposed columns are also nullable=false

Reply:
For this specific case, the "key" column must have nullable set to false. However, I do agree that we can optimize as a follow-up to infer nullability in general